<html><body><div><div class="post-text" itemprop="text">

<p>I'm having some trouble with some code for a djangp-reversion system I have set up.</p>

<p>I have a django model <code>ObjectClass</code> that inherits from a non-abstract model <code>_concept</code>. <sup>Long story, very unchangable now. Makes sense in the rest of the context of the code.</sup></p>

<p>On the class <code>_concept</code> I have mandatory ForeignKey field <code>workgroup</code> off to another model.</p>

<p>I have registered <code>ObjectClass</code> with Django reversion like so:</p>

<pre><code>reversion.unregister(ObjectClass)
reversion.register(ObjectClass,follow=['_concept_ptr'])
</code></pre>

<p>All old versions save and can be compared using 'django-reversion-compare'. Unfortunately when I then click on any old version to view the historical version I get the error:</p>

<pre><code>RevertError at /module.py
Could not revert revision, due to database integrity errors.
</code></pre>

<p>After some digging in the django-reversion code, I've done some fiddling and found that the error is coming up thus:</p>

<pre><code>RelatedObjectDoesNotExist at /module.py
_concept has no workgroup.
</code></pre>

<p>Now I've inspected the stored versions in the database and found a few things:</p>

<ol>
<li>Any given historical <code>reversion.models.version</code> of a _concept has a workgroup in the <code>serialized_data</code> field (which is expected).</li>
<li>Any given historical <code>reversion.models.version</code> of a ObjectClass <em>does not</em> have any of the parent information in the <code>serialized_data</code> field (which is expected).</li>
<li>Any given historical <code>reversion.models.version</code> of a ObjectClass <em>does not</em> have any a <code>_concept_ptr</code> in the <code>serialized_data</code> field (which is <strong>not</strong> expected).</li>
</ol>

<p>I'd suspect that <code>django-reversion</code> may have issues with fields that start with an underscore, <em>however</em> I have other fields that start with an underscore.</p>

<p>So I'm at a loss here. Is there a way to have this a model setup like this work?</p>

<hr/>

<p>edit:</p>

<p>After more checking it seems that the <code>has no workgroup</code> exception was from a Haystack call, which was alerting me to the fact that <code>reversion</code> is ignoring the workgroup for some reason.</p>

<p>I checked the database and this is whats in being serialised for an item (newlines added for readability):</p>

<pre><code>In [28]: myobj.serialized_data
Out[28]: u'[{"fields": {
               "definition": "&lt;p&gt;A code for sex.&lt;/p&gt;\\r\\n",
               "_is_locked": false, 
               "workgroup": 3, 
               "created": "2015-12-27T07:45:10.409Z", 
               "modified": "2015-12-27T08:38:26.989Z", 
               "readyToReview": false, 
               "_is_public": false, 
               "name": "Sex Code"
             }, 
             "model": "aristotle_mdr._concept", "pk": 30}]'
</code></pre>

<hr/>

<p>edit 2:</p>

<p>After disabling the haystack indexers everything works fine now, the issue is the Haystack signals are called when django-reversion tries to save the items to check consistency - then django calls the haystack <code>post_save</code> signals which try to update the index with incomplete data.</p>

<p>Still no solution yet. What I need in my haystack handler is either a way to determine if I'm inside a revisions transaction, or a way to prevent reversion from letting those signals fire. The latter is probably a better long term goal as I suspect that just by looking at revisions it is updating the Haystack index.</p>
    </div>
    </div></body></html>